architecture-guide.md

```diff
@@ -0,0 +1,350 @@
+  data and auxiliary controls interacting with component's API.
+  
+  It makes the pages simple, easy to maintain and directly presenting the API of
+  a given component.
+
+- Storybook pages are valuable for all types of components - both basic ones
+  (delegates, buttons) and more complex ones, doing complex flows like e.g.
+  funds transfer.
+
+- Storybook (and unit tests) has a mechanism of so-called `stubs` to replace
+  real stores with empty objects, but in a way preserving types. It's needed
+  because in Storybook and unit tests backend's context properties used in real
+  stores are not available. Store's stubs are intended to be empty QtObjects,
+  actual mocking is should be done within a page.
+  
+  This approach allows making the dependencies of the component truly visible
+  and explicit in a Storybook page. An alternative approach would be to
+  implement single, shared mocks for every store and share the among pages. But
+  it leads to situation that given mock provides much more than is needed by a
+  single UI component. As a consequence, it's not clear on which part of the
+  store given component depends on. A separate thing is that UI components
+  should not depend on the store at all unless it is really necessary (only top
+  level components covering complex flows).
+
+- A component in Storybook should be functional.
+  
+  The state of a component's Storybook page quite accurately reflects the state
+  of the component. Components with a well-designed API and no unnecessary
+  dependencies are easy to instantiate in isolation in Storybook. Conversely,
+  overly complex components are a nightmare in Storybook.
+
+- Use identified modules to import stores (using unquoted identifiers). It's
+  required for the stubs mechanism to work properly.
+
+  In practice imports like `import AppLayouts.Wallet.stores 1.0 as WalletStores`
+  should be used instead of `import "./stores" as WalletStores`. First version is
+  relative import from the file system. As a consequence, the mechanism for
+  overwriting import paths for tests and Storybook's pages will not work.
+
+
+## Code Style
+- QML code should be in-line with https://doc.qt.io/qt-5/qml-codingconventions.html
+- Top level component `id` should be always `root`
+- Private properties should be hidden in a `d` object (`QtObject { id: d }`, or
+  `QObject { id: d }`)
+- `objectName` property should not be specified for the top level component in
+  agiven file.
+  
+  Component may be used in various context and usually should have different
+  name in every context to disambiguate when doing lookup in unit tests, squish
+  tests or monitoring tool.
+- `qmldir`: entries should be sorted (`Alt+Shift+S` on selection in `QtCreator`)
+- Comments and documentation should cover parts which are not obvious, can be
+  skipped where intention is clear
+
+
+## Other
+- Cryptocurrency balances should be always handled as a big integer strings and
```

Comment:
Not sure if there is a convention for "big integers strings". How about using "hex encoded" prefixed with `0x` or `0X` to carry values that don't fit an QML `int`. And also in general wherever a big int would be used.

Reply:
This approach is currently quite commonly used, along with `AmountsArithmetic` utility for handling those string big ints. Using hex would mean big refactor and conversion in many places, probably also bit more problematic debugging because hex is not as direct and easy to read as decimal number.

Reply:
Also because, as we discovered recently, the Big.js library that we are using can construct numbers not only from integer strings, but also float strings, and even plain JS Number (with some precision lost ofc). This is very handy since you can use the same "big number" for doing the calculations but also for displaying to the user; the only thing one needs to do is to (optionally) convert the decimal separator back and forth

Reply:
So passing around values as strings between different layers (go/NIM/QML/JS) is possible:

```qml
(SwapModalAdaptor)

property string fromTokenAmount: "123456789.1234567899999999999"
```

and in the code, we can just call `var bigNum = AmountsArithmetic.fromString(fromTokenAmount)` and work with it as a number, pass it around to NIM w/o losing precision, or even display it to the user with code like `bigNum.toFixed().replace('.', LocaleUtils.userInputLocale.decimalPoint)` and we have a plain, localized string that can be pasted into an input field

Reply:
> ... Using hex would mean big refactor and conversion in many places ...

Then the strings should have decimal representation? Then it should be specified as such. In blockchain world hex strings with prefix are the norm.

> ... Using hex would mean big refactor and conversion in many places ...

My understanding is that all this string representations will pe passed to the Big lib before operations. Doesn't `Big.fromString` support hex parsing?

Reply:
The problem is that writing here that it should be hex will be dead remark, because we have tens of places where it's used in decimal form and none with hex. To introduce hex, everything would have to be refactored. If that's done for some reason, we can introduce the rule what's always hex.